lib/mix/tasks/oban.install.ex

```diff
@@ -0,0 +1,166 @@
+    ```
+
+    ## Options
+
+    * `--engine` or `-e` — Select the engine for your repo, defaults to `Oban.Engines.Postgres`
+    * `--notifier` or `-n` — Select the pubsub notifier, defaults to `Oban.Notifiers.Postgres`
+    * `--repo` or `-r` — Specify an Ecto repo for Oban to use
+    """
+  end
+end
+
+if Code.ensure_loaded?(Igniter) do
+  defmodule Mix.Tasks.Oban.Install do
+    @shortdoc __MODULE__.Docs.short_doc()
+
+    @moduledoc __MODULE__.Docs.long_doc()
+
+    use Igniter.Mix.Task
+
+    @impl Igniter.Mix.Task
+    def info(_argv, _composing_task) do
+      %Igniter.Mix.Task.Info{
+        group: :oban,
+        adds_deps: [oban: "oban"],
+        installs: [],
+        example: __MODULE__.Docs.example(),
+        only: nil,
+        positional: [],
+        composes: [],
+        schema: [engine: :string, notifier: :string, repo: :string],
+        defaults: [],
+        aliases: [e: :engine, n: :notifier, r: :repo],
+        required: []
+      }
+    end
+
+    @impl Igniter.Mix.Task
+    def igniter(igniter) do
+      app_name = Igniter.Project.Application.app_name(igniter)
+      opts = igniter.args.options
+
+      case extract_repo(igniter, app_name, opts[:repo]) do
+        {:ok, repo} ->
+          engine = parse_engine(repo, opts[:engine])
+          notifier = parse_notifier(repo, opts[:notifier])
+
+          conf_code = [engine: engine, notifier: notifier, queues: [default: 10], repo: repo]
+          test_code = [testing: :manual]
+
+          tree_code =
+            quote do
+              Application.fetch_env!(unquote(app_name), Oban)
+            end
+
+          migration = """
+          use Ecto.Migration
+
+          def up, do: Oban.Migration.up()
+
+          def down, do: Oban.Migration.down(version: 1)
+          """
+
+          igniter
+          |> Igniter.Project.Deps.add_dep({:oban, "~> 2.18"})
+          |> Igniter.Project.Config.configure("config.exs", app_name, [Oban], {:code, conf_code})
+          |> Igniter.Project.Config.configure("test.exs", app_name, [Oban], {:code, test_code})
+          |> Igniter.Project.Application.add_new_child({Oban, {:code, tree_code}}, after: [repo])
+          |> Igniter.Project.Formatter.import_dep(:oban)
+          |> Igniter.Libs.Ecto.gen_migration(repo, "add_oban", body: migration)
+
+        {:error, igniter} ->
+          igniter
+      end
+    end
+
+    defp extract_repo(igniter, app_name, nil) do
+      case Application.get_env(app_name, :ecto_repos, []) do
+        [repo | _] ->
+          {:ok, repo}
+
+        _ ->
+          issue = """
+          No ecto repos found for #{inspect(app_name)}.
+
+          Ensure `:ecto` is installed and configured for the current application.
+          """
+
+          {:error, Igniter.add_issue(igniter, issue)}
+      end
+    end
+
+    defp extract_repo(igniter, _app_name, module) do
+      repo = Igniter.Project.Module.parse(module)
+
+      if Code.ensure_loaded?(repo) do
```

Comment:
One of the things that can make installers much more composable is not using the compiled information about a module. For example, in `ash_postgres` we generate a repo for you if it doesn’t exist. So if someone says:

`mix igniter.install ash_postgres,oban`, this would fail because the module isn’t compiled.

to address that, you can use `Igniter.Project.Module.find_module` to look up the module. This gives you back a zipper. Then you can go to the `use Ecto.Repo` and look for the `adapter` option and use `Igniter.Code.Common.expand_literal` to get a proper module name out.


Along these lines, something I’ve considered doing is adding something like `Igniter.ensure_all_compiled` which will compile (but *not save*) everything that’s been done so far. It’s a big lift though so probably wouldn’t happen soon. If we did that then you could just drop it at the beginning of this task and then introspect modules to your hearts desire.

Anyway, the suggestion is to use the source code and not the compiled information to allow for composability with tools that generate the repo. Not strictly necessary, and could also be done later. I will probably PR some additional checks that make this installer work against `AshPostgres` and `AshSqlite` later if you use the repo source code, as those modify the `use` statement :)

Reply:
Good note. Ideally this would compose with modules that aren't compiled yet.

The initial approach was to use `Ecto.list_repos/1` to find the defaults and `Module.module_exists/2` to verify that the module was available. Each failed to find the expected modules, which could be down to the fact they're in `test/support/repo` and there are multiple modules in the same file. I'll give the `find_module` and parse approach a try.

Reply:
I'm building a big installer for new Ash projects, and would love to have a single checkbox for "Background Jobs" that installs both Oban and the Ash integration for it in one go❤️.

Reply:
I'll probably reach out to workshop the copy I use for the tooltip for that option, and how it links to Oban etc.

Reply:
Done!